text/0013-api-calls-and-fhir.md

```diff
@@ -0,0 +1,76 @@
+# API Calls and FHIR
+- Start Date: 2019/07/02
+- RFC PR: https://github.com/openmrs/openmrs-rfc-frontend/pull/13
+
+## Decision, including impact on distributions
+All API calls from the browser to the server go through the `@openmrs/api` javascript
+module.
+
+The `@openmrs/api` provides the following:
+- `authenticatedFetch`
+- `fhir`
+- `getCurrentPatient`, and a few similar `getCurrent...` functions for shared data.
+
+Whenever a FHIR backend API is available, it should be used. When no such FHIR api exists,
+the OpenMRS-specific APIs will be used.
+
+## Definition
+- `fhir` refers to the [official fhir.js npm package](https://github.com/FHIR/fhir.js/). [FHIR itself](https://www.hl7.org/fhir/overview.html)
```

Comment:
Have we considered using the java based HAPI FHIR APIs by James Agnew & team, https://github.com/jamesagnew/hapi-fhir
Their project seems to be more active than the fair.js one

Reply:
yes. We should evaluate which library to use. However, cautioning again on the common model, which i am sure will be difficult to agree to! (e.g. Patient attributes, would probably need specific extensions, or relationships etc) 
I am hoping that we use "Adapters"/"proxies" for FHIR as discussed before. So if the current patient fhir model (provided by openmrs), I should have ability to override that and fetch from a different api (as long as we agree to return a min common agreed model).

Reply:
Java doesn't work in the browser. Since HAPI is a java library, we can't use it in the browser.

The decision about HAPI is one for the backend, not the frontend.

Reply:
@angshu without a common model it's impossible to prevent duplicate network requests. I very much agree with you that the common model is troublesome and won't work for many use cases. For the code that doesn't want to use the common model, they can use `authenticatedFetch` or `fhir` to call whichever endpoint they want to however they want to (instead of `getCurrentPatient` which provides only the common model). That would lead in duplicate network requests, but is okay sometimes.